vllm/v1/struct_output/utils.py

```diff
@@ -0,0 +1,289 @@
+# SPDX-License-Identifier: Apache-2.0
+
+import json
+import re
+
+import xgrammar
+
+from vllm.sampling_params import SamplingParams
+
+
+def has_xgrammar_unsupported_json_features(schema: dict) -> bool:
+    """Check if JSON schema contains features unsupported by xgrammar."""
+
+    def check_object(obj: dict) -> bool:
+        if not isinstance(obj, dict):
+            return False
+
+        # Check for pattern restrictions
+        if "pattern" in obj:
+            return True
+
+        # Check for enum restrictions
+        if "enum" in obj:
+            return True
+
+        # Check for numeric ranges
+        if obj.get("type") in ("integer", "number") and any(
+                key in obj for key in [
+                    "minimum", "maximum", "exclusiveMinimum",
+                    "exclusiveMaximum", "multipleOf"
+                ]):
```

Comment:
nit: better to use tuples for these (same for others below)
```suggestion
                key in obj for key in (
                    "minimum", "maximum", "exclusiveMinimum",
                    "exclusiveMaximum", "multipleOf"
                )):
```